### PR TITLE
Updated BraveCompanions & FreshRecruits logic

### PR DESCRIPTION
**Pull request by Patane97**
Fixes a bug with Brave Companions (and Fresh Recruits) by updating its logic. It will now correctly highlight the valid cards based on the remaining _**Traits**_ when looking in the deck, even if a card with multiple of those _**Traits**_ is selected. For example, if you find a character with _**Commander. Mercenary.**_, they could be selected as the _**Commander**_ card or the _**Mercenary**_ card, meaning both traits are still available for selection.
This update uses permutation to find all valid combinations of the selected cards against the given traits, and returns the list of remaining traits that can still be selected.
Discord: Deathlysteve#4587 for any queries.

### CODE REVIEW

File: server/game/cards/11.1-TSC/FreshRecruits.js [hunk 1]
```diff
@@ -1,25 +1,56 @@
...
+                var next = current.slice();
+                next.push(r);
+                var newRemaining = remaining.slice();
+                newRemaining.splice(index, 1);
+                this.permutate(generated, next, newRemaining);
+            });
+        } else {
+            generated.push(current);
+        }
+        return generated;
```
Comment (ystros) [inline server/game/cards/11.1-TSC/FreshRecruits.js:53]: This logic is pretty hard to follow and is duplicated between both cards.

1. Is there something simpler that would have the same effect?
2. Can we improve the naming so that it is easier to follow? Particularly the numerous one-letter variables.
3. Can would put this logic in a common spot that both cards can use?
4. Some tests might be useful.

**Review by ystros — CHANGES_REQUESTED** [review]

File: server/game/cards/20-HMW/BraveCompanions.js [hunk 2]
```diff
@@ -40,7 +40,46 @@ class BraveCompanions extends DrawCard {
...
+        for(let card of cards) {
+            player.moveCard(card, 'hand');
+        }
+        this.game.addMessage('{0} uses {1} to search their deck and adds {2} to their hand', player, this, cards);
+        return true;
+    }
+
+    cancelSearch(player) {
+        this.game.addMessage('{0} uses {1} to search their deck, but does not add any cards to their hand', player, this);
+        return true;
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/cards/20-HMW/BraveCompanions.js:82]: Yep, these were pasted from the play-testing repository (where it wasnt using GameActions). Removed+resolved.

File: server/game/cards/11.1-TSC/FreshRecruits.js [hunk 3]
```diff
@@ -1,25 +1,56 @@
...
     setupCardAbilities() {
         this.action({
             title: 'Search deck',
-            handler: () => {
-                this.game.promptForDeckSearch(this.controller, {
-                    activePromptTitle: 'Select cards',
-                    numToSelect: 3,
-                    cardCondition: (card, context) => card.getType() === 'character' && this.remainingTraits(context.selectedCards).some(trait => card.hasTrait(trait)),
-                    onSelect: (player, cards) => this.selectCards(player, cards),
-                    onCancel: player => this.cancelSearch(player),
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/cards/11.1-TSC/FreshRecruits.js:13]: Yep! Removed+Resolved.

File: server/game/cards/11.1-TSC/FreshRecruits.js [hunk 4]
```diff
@@ -1,25 +1,56 @@
...
     }
 
     remainingTraits(selectedCards) {
         const traits = ['Ranger', 'Builder', 'Steward'];
-        return traits.filter(trait => !selectedCards.some(card => card.hasTrait(trait)));
+        let tempCards = selectedCards.slice();
+        for(let i = tempCards.length ; i < traits.length ; i++) {
+            tempCards.push(null);
+        }
+        var remaining = [];
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/cards/11.1-TSC/FreshRecruits.js:31]: Changed from `var` to `let` on both FreshRecruits + BraveCompanions.

File: server/game/cards/11.1-TSC/FreshRecruits.js (hunk 1 above)
Comment (Patane97) [reply server/game/cards/11.1-TSC/FreshRecruits.js:53]: 1. I was working on this for a few days straight looking for a simpler solution (outside of selecting all 3 traits in a separate prompt). Not saying there isn't a simpler solution out there, just not sure I was able to find it. :)
2. Happy to update the 1 letter variables for clarity.
3. I was actually looking for an appropriate place to put this but couldnt find one quickly. I have designed it to work for any combination of traits (eg, if there is a future card that searches for a _**Companion**_, a _**Lord**_, a _**Lady**_ and a _**Bastard**_ character, this could do it by just adjusting the Traits constant. If you could point me to the right class to put this method in, id happily move/call it from there :)
4. I'll look into how the tests work as I am unfamiliar with it at this time. Any suggestions on how to proceed with tests?

**Comment by EvadLhorg** [comment]
relates to #3142

File: server/game/cards/11.1-TSC/FreshRecruits.js (hunk 1 above)
Comment (ystros) [reply server/game/cards/11.1-TSC/FreshRecruits.js:53]: 1. I haven't been able to able to devote a lot of time, but I also was trying to find a simpler solution and couldn't quite get there. I think comments about the various stages might be useful, as the code is somewhat dense (even if naming is updated).
3. I'm not sure there's an existing spot that would fit well. I wouldn't want to add the method to the Game, Player, or Card classes. I've toyed with the idea of introducing a `CardConditions` class/file that would include helpers for common things (e.g. "all controlled cards are the same faction", "there's a plot revealed with a specific trait", etc). Perhaps create that as a new file and add the method that gives you the remaining traits?
4. I'd say either write tests directly for your new method (so you'd add it to a file with the same name but with `.spec.js` at the end instead) that just test various input-output combinations, or you could add an integration test that starts a game, activates either Brave Companions or Fresh Recruits, and tries selecting cards. Example integration test: https://github.com/throneteki/throneteki/blob/master/test/server/cards/15-DotE/ArchmaesterMarwyn.spec.js. If you can't figure out, don't worry about it too much, I can probably create one for you if needed.

Comment (Patane97) [reply server/game/cards/11.1-TSC/FreshRecruits.js]: I believe I have resolved these issues (tried to leave as much commenting as possible to explain whats going on). I have not attempted card testing for this yet, but will work on that next. Up to you, but this should be ready to push if youre happy! :)

**Review by ystros — APPROVED** [review]
